### ISSUE TITLE
Feature Proposal: multi-optimizer / multi-transform

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#129]
Currently, to apply different optimizers to different sets of parameters, you need to construct multiple masks and chain them. For example, to optionally apply weight decay to some parameters and not to others:

```py
tx = optax.chain(
    optax.masked(optax.adamw(0.01, weight_decay=0.0), 
                 mask=partial(jax.tree_map, lambda p: p.ndim == 1)),
    optax.masked(optax.adamw(0.01, weight_decay=1e-3), 
                 mask=partial(jax.tree_map, lambda p: p.ndim != 1)),
)
```

However, this has a few issues:
1. When you have more than two groups, creating mask functions for them all becomes verbose
2. The burden of verifying the masked groups are mutually exclusive falls on the user (which can be bothersome for groups > 2)
3. This solution isn't obvious to a new user

I'd like to propose a `multi_transform` (or `multi_transformation`, as always naming is hard) that generalizes `optax.masked`:

```py
def multi_transform(transforms: Sequence[GradientTransformation], 
                    partition: Union[PyTree, Callable[[base.Params], PyTree]]):
  ... 
```
Where `partition` is a pytree where the leaves contain an index (or a function that returns such a pytree given the parameters). The index corresponds to a `GradientTransformation` in `transforms`.

With this, our above example would be:
```py
tx = optax.multi_transform(
    [optax.adamw(0.01, weight_decay=1e-3), optax.adamw(0.01, weight_decay=0.0)],
    partition=partial(jax.tree_map, lambda p: int(p.ndim == 1)))  # 0 for weight decay, 1 for no weight decay
```

This solves our above problems:
1. You only need one pytree (function)
2. The groups are mutually exclusive by design, since there's only one pytree with indices
3. The name matches what users would expect from such a feature (due to similar naming in other libraries, e.g. [Flax](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.MultiOptimizer)).

Minor details:
- `partition` could just be `partition_fn: Callable[[base.Params], PyTree]` (I left it as `partition` to mirror the signature of `masked`).
- `partition` could have a different name, perhaps `indices`.

cc: @mtthss @jheek @andsteing

**Comment by avital** [comment]
I agree that it's hard to make sure that masks don't overlap thus I agree with the motivation here.

But I am concerned about the proposed API because it means you have to keep track of integer indices in the first argument to `multi_transform`. And if someone adds another optimizer somewhere in the list (not at the end) then things will silently fail.

Can there be an API that looks somehow more like the old flax MultiOptim? Where you directly couple a "mask" to an Optax optimizer, rather than going through indices?

```py
optax.multi_transform(
  partition=jax.tree_map, lambda p: optax.adamw(0.01, weight_decay=0.0) if p.ndim == 1) else

**Comment by n2cholas** [comment]
Good point @avital, I like that idea. I think the example in my post is a bit too simple, here's something more realistic that illustrates the issue and your proposal for new readers (assuming we use Flax Conv and BatchNorm layers):

```py
def map_params(fn):  # Convenience function to construct mask_fns
    def map_fn(params):
        flat_params = flax.traverse_util.flatten_dict(params)
        flat_mask = {path: fn(path) for path in flat_params.keys()}
        return flax.traverse_util.unflatten_dict(flat_mask)
    return map_fn

# Suppose we want different learning rates for kernels, biases, and all Batchnorm params
# Below implementation has a silent bug: both (b) and (c) will be applied to batchnorm biases
tx = optax.chain(
    optax.masked(optax.sgd(0.1), map_params(lambda path: path[-1] == 'kernel')),  # (a)
    optax.masked(optax.sgd(0.2), map_params(lambda path: path[-1] == 'bias')),  # (b)
    optax.masked(optax.sgd(0.3), map_params(lambda path: path[-2].startswith('BatchNorm'))),  # (c)
)

# Original Proposal:
# It is _impossible_ to have overlap by design, but using indices is brittle
def partition_fn(path):
    if path[-2].startswith('BatchNorm'): return 0
    if path[-1] == 'kernel': return 1
    if path[-1] == 'bias': return 2

tx = optax.multi_transform([optax.sgd(0.1), optax.sgd(0.2), optax.sgd(0.3)], 
                           map_params(partition_fn))

# Avital's Proposal:
# This will throw an error, since multi_transform will check for overlap.
tx = optax.multi_transform(
    (optax.sgd(0.1), map_params(lambda path: path[-1] == 'kernel')),  # (a)
    (optax.sgd(0.2), map_params(lambda path: path[-1] == 'bias')),  # (b)
    (optax.sgd(0.3), map_params(lambda path: path[-2] == 'BatchNorm')),  # (c)
)
# User would correct to something like this
tx = optax.multi_transform(
    (optax.sgd(0.1), map_params(lambda path: path[-1] == 'kernel'),  # (a)
    (optax.sgd(0.2), map_params(lambda path: path[-1] == 'bias' and not path[-2].startswith('BatchNorm')),  # (b)
    (optax.sgd(0.3), map_params(lambda path: path[-2].startswith('BatchNorm')),  # (c)
)
```

Should the `multi_transform` check that users provide a transform for all parameters? i.e., check that the union of all the masks covers all parameters. This protects users against the opposite of the overlap bug: not having a transform for some parameters. If users truly want to skip optimizing some parameters, they can always use `optax.masked(optax.multi_transform(...))`.

**Comment by andsteing** [comment]
Another simple modification to your original proposal would be to simply use `dict` instead of `Sequence` when providing the individual gradient transformations, and then to refer to them via their key, i.e.

```python
# Original Proposal - modified:
# It is _impossible_ to have overlap by design, but using indices is brittle
def partition_fn(path):
    if path[-2].startswith('BatchNorm'): return 'batch_norm'
    if path[-1] == 'kernel': return 'kernel'
    if path[-1] == 'bias': return 'bias'

tx = optax.multi_transform(dict(batch_norm=optax.sgd(0.3),
                                kernel=optax.sgd(0.1),
                                bias=optax.sgd(0.2)), 
                           map_params(partition_fn))
```

**Comment by mtthss** [comment]
Thanks @n2cholas for kicking off the discussion!
I like the version by @andsteing where the set of transforms is a dict
and the partition function returns the keys

In particular I like that:
1) it makes it clear that the order doesn't have a meaning
2) keys feel less brittle than indices

**Comment by n2cholas** [comment]
I like @andsteing's version as well! 

Question on naming: what's a better name than `partition` for the second argument? My issue with partition is in other Python libraries, partition is a function that takes one collection and returns two (e.g. `itertools.partition`, [`haiku.data_structures.partition`](https://dm-haiku.readthedocs.io/en/latest/api.html#haiku.data_structures.partition)). In our case, it's a function that simply "labels" each parameter (or subtree), so that may be confusing to a new user.

Perhaps a better name is `label_fn` or `label`? Though that also has a more common meaning in ML. Any ideas?
